api/internal/handler/consumer/consumer.go

```diff
@@ -147,11 +147,12 @@ func (h *Handler) Set(c droplet.Context) (interface{}, error) {
 	input.Consumer.ID = input.Consumer.Username
 	ensurePluginsDefValue(input.Plugins)
 
-	if err := h.consumerStore.Update(c.Context(), &input.Consumer, true); err != nil {
+	ret, err := h.consumerStore.Update(c.Context(), &input.Consumer, true)
+	if err != nil {
 		return handler.SpecCodeResponse(err), err
 	}
 
-	return nil, nil
+	return ret, nil
```

Comment:
We can directly return `&input.Consumer`, because store is just return the pointer of it.
So that we can avoid to return object in `Store`, keep one way to get changes would be better.

Reply:
ping @imjoey

Reply:
@ShiningRush  Thanks for your reviewing. IMHO, directly returning the `&input.Consumer` would mix up the DTO and the DAO, which may be confusing to developer. Please allow me use the following case as an example:

```go
type SetInput struct {
	entity.Consumer
	Username string `auto_read:"username,path"`
}

func (h *Handler) Set(c droplet.Context) (interface{}, error) {
	input := c.Input().(*SetInput)
	if input.ID != nil && utils.InterfaceToString(input.ID) != input.Username {
		return &data.SpecCodeResponse{StatusCode: http.StatusBadRequest},
			fmt.Errorf("consumer's id and username must be a same value")
	}
	if input.Username != "" {
		input.Consumer.Username = input.Username
	}
	input.Consumer.ID = input.Consumer.Username
	ensurePluginsDefValue(input.Plugins)

	err := h.consumerStore.Update(c.Context(), &input.Consumer, true)
	if err != nil {
		return handler.SpecCodeResponse(err), err
	}

        // Notice: Desensitize the response data  =======
        input.Consumer.Desc = "*******"

	return &input.Consumer, nil
}

```

As the comment mentioned above, in order to do customization before sending back to client, we need to change input value for output, which seems some sort of werid.

Looking forward to your insights. Thanks.

Reply:
> directly returning the &input.Consumer would mix up the DTO and the DAO

In this case, input is not only `DTO` also a resource.
I support to return resource in `Patch` because it is indeed different with input.But `Update` method required a complete resource, so DTO also is a resource.

> which may be confusing to developer

On the contrary, the situation that confuses developers is that you return an object and make them think that you did not perform any operations on the original object, but this is not the truth.

BTW: here is a better way to implement you example that move logic to entity,so code could be
```go
....

input.Consumer.Desensitize()
return &input.Consumer, nil
```
`rich domain model` would be more meaningful.

Reply:
@ShiningRush 

Thanks for getting back. Let's wait for others opinions. 😄 

> BTW: here is a better way to implement you example that move logic to entity,so code could be
Thanks. The code you provided is awesome. While IMHO, it is still not able to resolve the mixup of concepts of request and response.